sample/weighted.go

```diff
@@ -0,0 +1,98 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import "math/rand"
+
+// Weighted provides sampling without replacement from a collection of items with
+// weighted probabilites.
+type Weighted struct {
+	weights []float64
+	heap    []float64
```

Comment:
I like the name update. I still think it deserves an extended comment about the data structure. In general Go is not tricky with data structures, but here we are being tricky. It's for a good cause, but since it's more complicated than it appears it deserves a comment. If there's a good description somewhere in the standard library, it's fine to link it, otherwise I'd appreciate seeing something along the lines of:

````
heap accumulates the probabilities using an in-memory binary heap to allow sampling in log(n) time rather than linear.

Each element, i, in the heap contains the weight of element i plus the sum of the weights of the nodes below it. The root of the heap (index zero) thus contains the full some of the weights. The heap is traversed by [indexing method to travel to the childen and indexing method to travel to the parent]
````